src/fetchers/tarball-fetcher.js

```diff
@@ -146,7 +147,10 @@ export default class TarballFetcher extends BaseFetcher {
       this.validateError = err;
     });
     integrityValidateStream.once('error', err => {
-      this.validateError = err;
+      if (this.validateError == null) {
+        // Don't overwrite error if one is already there
```

Comment:
A quick-fix for the logic here is to just use separate variables for errors from the two streams.

The more general error chaining issue is more tricky here, and it would help if someone with with more knowledge about streams can help out here. If for example `integrityValidateStream` emits an error, `untarStream` still tries to continue parsing, and emits it's own error. So for the specific case of a truncated response, we actually end up with three errors: Content-Length mismatch, hash mismatch and "unexpected end of file". The "unexpected end of file" is the last one in the chain, but is the one that actually ends up being passed on to the user.

I would have expected `stream.pipe` to pass on errors as well, but I may be misunderstanding how streams work in Node. So not sure what the cleanest way is to handle this.

Reply:
As far as I know, this is the way to collect errors with streams. For better readability, I might go for something like `stream1.on('error', handleError).pipe(stream2).on('error', handleError)`, but I think in this case that would involve quite a bit of shifting around...